azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+        assert threading.current_thread().name == "callback"
+        callback_called.set()
+
+    stage.on_connected_handler = callback
+
+    stage.on_connected()
+    callback_called.wait()
+
+
+@pytest.mark.it("Calls on_disconnected_handler in callback thread")
+def _test_pipeline_root_runs_on_disconnected_in_callback_thread(self, stage, mocker):
+    stage.pipeline_root = stage
+    callback_called = threading.Event()
+
+    def callback(*arg, **argv):
+        assert threading.current_thread().name == "callback"
+        callback_called.set()
+
+    stage.on_disconnected_handler = callback
+
+    stage.on_disconnected()
+    callback_called.wait()
+
+
+@pytest.mark.it("Calls on_event_received_handler in callback thread")
+def _test_pipeline_root_runs_on_event_received_in_callback_thread(self, stage, mocker, event):
+    stage.pipeline_root = stage
+    callback_called = threading.Event()
+
+    def callback(*arg, **argv):
+        assert threading.current_thread().name == "callback"
+        callback_called.set()
+
+    stage.on_pipeline_event_handler = callback
+
+    stage.handle_pipeline_event(event)
+    callback_called.wait()
+
+
 TestPipelineRootStagePipelineThreading.test_runs_callback_in_callback_thread = (
     _test_pipeline_root_runs_callback_in_callback_thread
 )
 TestPipelineRootStagePipelineThreading.test_runs_operation_in_pipeline_thread = (
     _test_pipeline_root_runs_operation_in_pipeline_thread
 )
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_connected_in_callback_thread = (
+    _test_pipeline_root_runs_on_connected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_disconnected_in_callback_thread = (
+    _test_pipeline_root_runs_on_disconnected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_event_received_in_callback_thread = (
+    _test_pipeline_root_runs_on_event_received_in_callback_thread
+)
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.EnsureConnectionStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_mqtt.MQTTPublishOperation,
+        pipeline_ops_mqtt.MQTTSubscribeOperation,
+        pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+)
+
+fake_topic = "__fake_topic__"
+fake_payload = "__fake_payload__"
+ops_that_cause_connection = [
+    {
+        "op_class": pipeline_ops_mqtt.MQTTPublishOperation,
+        "op_init_kwargs": {"topic": fake_topic, "payload": fake_payload},
+    },
+    {"op_class": pipeline_ops_mqtt.MQTTSubscribeOperation, "op_init_kwargs": {"topic": fake_topic}},
+    {
+        "op_class": pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+        "op_init_kwargs": {"topic": fake_topic},
+    },
+]
+
+
+@pytest.mark.parametrize(
+    "params",
+    ops_that_cause_connection,
+    ids=[x["op_class"].__name__ for x in ops_that_cause_connection],
```

Comment:
What about all the other operations that could pass through this stage?

Reply:
Yeah, what about them?  What do you mean?  add_base_pipeline_stage_tests() makes sure they pass through.

---
In reply to: [312665049](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312665049) [](ancestors = 312665049)